man/man1/obd2gpx.1
.TH obd2gpx 1
.SH NAME
obd2gpx \- Convert obdgpslogger(1) logs to gpx files

.SH SYNOPSIS
.B obd2gpx [ options ]

.SH DESCRIPTION
.IX Header "DESCRIPTION"
Convert obdgpslogger(1) logs to gpx files

.SH OPTIONS
.IX Header "OPTIONS"
.IP "-o|--out <output filename>"
Output to this .gpx file
.IP "-d|--db <database>"
Work from logs stored in this database file
.IP "-v|--version"
Print out version number and exit.
.IP "-h|--help"
Print out help and exit.
 
.SH SEE ALSO
.IX Header "SEE ALSO"
.BR "obdgpslogger(1), obd2gpx(1), obd2kml(1), obdsim(1), obdgui(1), obdlogrepair(1)"

.SH AUTHORS
Gary "Chunky Ks" Briggs <chunky@icculus.org>



ChangeLog
Version 0.16 (released 2011-05-26)
 comm: Much longer timeout for devices that take a long time to init
 kml: minor performance and SQL cosmetic fixes
 logger: fix long-standing bug when gps sometimes appears to lock up
 Initial add of livekml PHP script for watching live Google Earth
 Conf: Accept hex PIDs as well as human names for log_columns
 Sim Cycle gen: provide pids up to 0x60
 build: Option to use systemwide sqlite3 lib instead of local copy
 Sim: Support protocols with ATSP{A,}{0-9,A-C}, show appropriate headers
 Sim: Add --benchmark option for occasional updates
 Sim: Refactor to simplify main loop
 Sim: Add a timeout for each individual ecu
 Sim: Add simple checksum when headers are enabled
 Sim: ATD{0,1} support

Version 0.15 (released 2010-06-18)
 Sim: Always return errors in groups of three, padded if necessary
 Logrepair: split repair functions into separate library
 kml: Fix errors that caused some singlevalue plots to not plot
 Sim: Freezeframe [mode 0x02] support
 Logrepair: run PRAGMA integrity_check
 Arch Linux packaging
 Initial add of obd2gpx converter
 kml: Add graph based solely on gps data
 Logger: log course,speed,gpstime from gps

Version 0.14 (released 2010-04-02)
 Sim: gui_fltk has dials instead of input boxes
 obdfft: Show axes ranges on graph, fix CSV parsing issue
 Logger: Bundle up several seconds of updates into single transactions
 Sim: Option to listen on bluetooth
 Logger: Baudrate guesser
 Logger: Baudrate upgrader
 Move obd comms into separate library
 Sim: Multiple ECU support [attach a generator for each ecu]
 Sim: Random,cycle generators don't return values they can't support
 Sim: Support for modes 0x03,0x04
 Sim: Add "Error" data generator
 Sim: gui_fltk now has support for DTCs

Version 0.13 (released 2010-02-27)
 Opensolaris packaging [for jucr]
 Read OBD_CONFIGFILE environment variable for a config file
 Initial add of simple data analysis tool [fourier filter against db columns]
 Logger can redirect stdout/err to a file
 obdsim: Ability to use a real device instead of a pty on unix
 Logger: Add columns to log if tables already existed
 Logger: Create DB with IF NOT EXISTS to reduce warnings
 Logger,Logrepair: Add indexed tripid column to obd and gps tables
 Kml: Use tripids where possible instead of times to key on trips
 Kml: Warn if tripids are missing from obd or gps tables
 Kml: Don't output kml for trips without gps co-ordinates
 Kml: Add trip number to graphs
 Initial add of simple trip comparison tool [mpg, fuel usage]
 Sim: dlopen generator takes a sub-seed
 Logger: Gracefully handle delayed gpsd startup
 Logrepair: Run ANALYZE against log

Version 0.12 (released 2010-01-03)
 ftdipty: Support user-specified device and vendor ids
 Logger: Support gpsd V3 protocol
 Sim: Support ATI, ATL{0,1}
 Sim: Tidy up linefeed handling throughout app

Version 0.11 (released 2009-11-25)
 Logger: Recognise multi-line serial output, break out line parsing
 Logger: Work on reading error codes
 Logger: Always clean up serial port on early exit
 Sim: bugfixes
 Sim: Move to stdlib read/write instead of stdio
 Sim: Simplify output mechanism
 Sim: Use /dev/ptmx if posix_openpt isn't available
 Sim: Optionally attach a screen session
 Sim: Add "Cycle" generator
 Sim: Add longdesc to generators so they can self-document
 Docs: Add manpage for obdsim-dbus
 Sim: Convert to C++. Make simport abstract class
 Sim: Implement posix, windows simport classes
 Logger: Don't offer to daemonise on windows
 Sim: Support logging all serial comms
 Sim: Allow command-line options for faking AT{Z,@1}
 Logger: Attempt to guess supported PIDs if car reports zero for 0100
 Sim: Add shell of megasquirt code

Version 0.10 (released 2009-10-31)
 Logger, conf: Ability to adjust serial baudrate
 Initial add of ftdi-pty proxy for systems without ftdi-sio
 conf: Add ability to write config file back to disk
 ftdipty can modify config file for obdgpslogger
 Logger, conf: Ability to configure logfile name
 Logger: Gracefully exit if a serial write fails
 Logger,doc: Hotplug docs
 Logger: cygwin compilation fixes
 Sim: Don't build socket generator without getaddrinfo symbol
 Logger: If obd device isn't available, exit [would log gps before]

Version 0.9 (released 2009-09-25)
 Logger: Create indices on tables when the tables are created
 obd2kml: Small table optimisations
 New tool: obdlogrepair. Small tool to fix any minor issues in logfile
 obdsim: Initial add of socket generator [for wifi obd devices]
 obd2csv: Take --start and --end time parameters
 obd2csv: add option to gzip csv file
 obdsim: dbus plugin work with more types on the bus
 config,logger: Make columns to log run-time configurable
 obd2kml: Add date to start & end beacons
 Logger: fix stupid wait on startup
 Logger: Adding ability to handle OBD errors

Version 0.8 (released 2009-09-04)
 Put Copyright/GPL notices in all files
 Cosmetic build improvements
 obdsim: Add idle callback to generators
 obdsim: Messing with serial comms
 logger: Make newline configurable
 logger: Add timeout on database in case someone's scraping it
 obdsim: Initial add of fltk gui generator
 logger: Add option to daemonise
 obdsim: Add mode to all getvalue module calls
 logger: Catch SIGTERM to gracefully exit
 obdsim: Seed Random generator with passed seed
 manpage: describe obdsim seeds
 obd2csv,obd2kml,obdsim logger: Don't create the database file
 logger: adding some dbus methods

Version 0.7 (released 2009-06-25)
 Initial add of new module, obdsim
 obdsim plugins: dlopen, Logger, DBus, Random

Version 0.6 (released 2009-06-15)
 Move obdlogger source into its own subdir
 logger optimisation: remove all spaces in obd streams
 logger: add a serial port communications log
 logger: move capabilities into its own file, add features
 logger: Only create database columns, and only try to log, supported pids
 logger: add serial port logging option
 cmake: default gps dev [for ui] separately compile-time configurable

src/gpx/obd2gpx.c
/* Copyright 2009 Gary Briggs

This file is part of obdgpslogger.

obdgpslogger is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.

obdgpslogger is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with obdgpslogger.  If not, see <http://www.gnu.org/licenses/>.
*/


/** \file
 \brief obd2gpx main entrypoint
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <libgen.h>

#include "obd2gpx.h"
#include "obdconfig.h"

#include "sqlite3.h"

int main(int argc, char **argv) {
	/// Output file
	FILE *outfile;

	/// Database to dump
	sqlite3 *db;

	/// outfile filename
	char *outfilename = strdup(DEFAULT_OUTFILENAME);

	/// Database file to open
	char *databasename = strdup(OBD_DEFAULT_DATABASE);

	/// getopt's current option
	int optc;

	/// might get set during option parsing. Exit when done parsing
	int mustexit = 0;

	while ((optc = getopt_long (argc, argv, gpxshortopts, gpxlongopts, NULL)) != -1) {
		switch (optc) {
			case 'h':
				gpxprinthelp(argv[0]);
				mustexit = 1;
				break;
			case 'v':
				gpxprintversion();
				mustexit = 1;
				break;
			case 'd':
				if(NULL != databasename) {
					free(databasename);
				}
				databasename = strdup(optarg);
				break;
			case 'o':
				if(NULL != outfilename) {
					free(outfilename);
				}
				outfilename = strdup(optarg);
				break;
			default:
				gpxprinthelp(argv[0]);
				mustexit = 1;
				break;
		}
	}
	if(mustexit) exit(0);

	int rc;
	rc = sqlite3_open_v2(databasename, &db, SQLITE_OPEN_READONLY, NULL);
	if( SQLITE_OK != rc ) {
		fprintf(stderr, "Can't open database %s: %s\n", databasename, sqlite3_errmsg(db));
		sqlite3_close(db);
		exit(1);
	}

	char select_sql[4096] = "SELECT lat,lon,alt,time,trip FROM gps WHERE trip IS NOT NULL ORDER BY trip,time";

	sqlite3_stmt *select_stmt; // Our actual select statement
	
	rc = sqlite3_prepare_v2(db, select_sql, -1, &select_stmt, NULL);

	if(SQLITE_OK != rc) {
		printf("Error attempting to prepare select(%i): %s\n", rc, sqlite3_errmsg(db));
		sqlite3_close(db);
		exit(1);
	}


	if(NULL == (outfile = fopen(outfilename, "w"))) {
		perror(outfilename);
		sqlite3_close(db);
		exit(1);
	}

	gpx_writeheader(outfile, basename(outfilename));

	int currtrip = -1;

	while(SQLITE_ROW == sqlite3_step(select_stmt)) {
		double lat = sqlite3_column_double(select_stmt, 0);
		double lon = sqlite3_column_double(select_stmt, 1);
		double alt = sqlite3_column_double(select_stmt, 2);
		time_t datatime = sqlite3_column_int64(select_stmt, 3);
		int trip = sqlite3_column_int64(select_stmt, 4);

		if(currtrip != trip) {
			if(currtrip > -1) gpx_endtrip(outfile);
			printf("Writing trip %i\n", trip);
			gpx_starttrip(outfile, trip);
		}

		fprintf(outfile, "\t\t\t<trkpt lat=\"%f\" lon=\"%f\">\n", lat, lon);
		if(alt > -900) { // -1000 is a crappy way to tell if we had a 3d fix
			fprintf(outfile, "\t\t\t\t<ele>%f</ele>\n", alt);
			fprintf(outfile, "\t\t\t\t<fix>3d</fix>\n");
		} else {
			fprintf(outfile, "\t\t\t\t<fix>2d</fix>\n");
		}

		// xsd:dateTime format: CCYY-MM-DDThh:mm:ss 
		char timestring[128] = "\0";
		struct tm *tostr = localtime(&datatime);
		if(0 < strftime(timestring, sizeof(timestring), "%FT%T", tostr)) {
			fprintf(outfile, "\t\t\t\t<time>%s</time>\n", timestring);
		}

		fprintf(outfile, "\t\t\t</trkpt>\n");
		currtrip = trip;
	}
	gpx_endtrip(outfile);

	gpx_writetail(outfile);
	sqlite3_finalize(select_stmt);

	fclose(outfile);

	sqlite3_close(db);

	free(outfilename);
	free(databasename);

	return 0;
}

void gpx_writeheader(FILE *outfile, const char *filename) {
	fprintf(outfile, "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
			"<gpx version=\"1.1\" creator=\"obdgpslogger\"\n"
			"\t\txmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\n"
			"\t\txmlns=\"http://www.topografix.com/GPX/1.1\"\n"
			"\t\txsi:schemaLocation=\"http://www.topografix.com/GPS/1/1\n"
			"\t\thttp://www.topografix.com/GPX/1/1/gpx.xsd\">\n"
			"\t<metadata>\n"
			"\t\t<name>obd2gpx %s</name>\n"
			"\t\t<desc>OBDGPSLogger obd2gpx convert from %s</desc>\n"
			"\t\t<keywords>obdgpslogger,obd2gpx</keywords>\n"
			"\t\t<author>\n"
			"\t\t\t<name>Gary Briggs</name>\n"
			"\t\t\t<email id=\"chunky\" domain=\"icculus.org\" />\n"
			"\t\t\t<link href=\"http://icculus.org/obdgpslogger\"><text>OBDGPSLogger</text></link>\n"
			"\t\t</author>\n"
			"\t</metadata>\n", filename, filename);
}

void gpx_writetail(FILE *outfile) {
	fprintf(outfile, "</gpx>\n");
}

void gpx_starttrip(FILE *outfile, int tripnum) {
	fprintf(outfile, "\t<trk>\n"
		"\t\t<name>Trip %i</name>\n"
		"\t\t<trkseg>\n", tripnum);
}

void gpx_endtrip(FILE *outfile) {
	fprintf(outfile, "\t\t</trkseg>\n"
		"\t</trk>\n");
}

void gpxprinthelp(const char *argv0) {
	printf("Usage: %s [params]\n"
		"   [-o|--out=<" DEFAULT_OUTFILENAME ">]\n"
		"   [-d|--db=<" OBD_DEFAULT_DATABASE ">]\n"
		"   [-v|--version] [-h|--help]\n", argv0);
}

void gpxprintversion() {
	printf("Version: %i.%i\n", OBDGPSLOGGER_MAJOR_VERSION, OBDGPSLOGGER_MINOR_VERSION);
}

